src/outputs/parthenon_hdf5.cpp

```diff
@@ -15,17 +15,193 @@
+    char_ptrs[i] = values[i].c_str();
+  }
+  HDF5WriteAttribute(name, char_ptrs, location);
+}
 
-#ifdef HDF5OUTPUT
+template <>
+std::vector<std::string> HDF5ReadAttributeVec(hid_t location, const std::string &name) {
+  // get strings as char pointers, HDF5 will allocate the memory and we need to free it
+  auto char_ptrs = HDF5ReadAttributeVec<char *>(location, name);
 
-namespace parthenon {
+  // make strings out of char pointers, which copies the memory and then free the memeory
+  std::vector<std::string> res(char_ptrs.size());
+  for (size_t i = 0; i < res.size(); ++i) {
+    res[i] = std::string(char_ptrs[i]);
+    free(char_ptrs[i]);
+  }
+
+  return res;
+}
+
+// template specialization for bool
+template <>
+void HDF5WriteAttribute(const std::string &name, const std::vector<bool> &values,
+                        hid_t location) {
+  // can't use std::vector here because std::vector<bool>  doesn't have .data() member
+  std::unique_ptr<hbool_t[]> data(new hbool_t[values.size()]);
+  for (size_t i = 0; i < values.size(); ++i) {
+    data[i] = values[i];
+  }
+  HDF5WriteAttribute(name, values.size(), data.get(), location);
+}
+
+} // namespace HDF5
+
+namespace {
+
+std::string PackStrings(const std::vector<std::string> &strs, char delimiter) {
+  std::string pack;
+  for (const auto &s : strs) {
+    pack += s + delimiter;
+  }
+  return pack;
+}
+
+std::vector<std::string> UnpackStrings(const std::string &pack, char delimiter) {
+  std::vector<std::string> unpack;
+  const char *curr = pack.data();
+  const char *const end = curr + pack.size();
+
+  while (curr < end) {
+    const auto tab = strchr(curr, delimiter);
+    if (tab == nullptr) {
+      std::stringstream msg;
+      msg << "### ERROR: Pack string does not end with delimiter" << std::endl;
+      PARTHENON_FAIL(msg);
+    }
+
+    if (tab == curr) {
+      unpack.push_back("");
+    } else {
+      unpack.push_back(std::string(curr, tab - curr));
+      curr = tab + 1;
+    }
+  }
+
+  return unpack;
+}
+
+} // anonymous namespace
+
+using namespace HDF5;
+
+// Helper struct containing some information about a variable that we can easily
+// communicate via MPI
+struct VarInfo {
+  // We need to communicate this struct via MPI. To Make our lives a bit easier, we will
+  // combine the vlen integer and the is_sparse and is_vector flags into a single int
+  // (call it info_code) and communicate that.
+  //
+  // The info_code will have the vlen in the lower 16 bits and bits 20 and 21 will encode
+  // the is_sparse and is_vector flags
+  static constexpr int max_vlen = (1 << 16) - 1;
+  static constexpr int sparse_flag = (1 << 20);
+  static constexpr int vector_flag = (1 << 21);
+
+  std::string label;
+  std::vector<std::string> component_labels; // may be empty
+  int vlen;
+  bool is_sparse;
+  bool is_vector;
+
+ private:
+  VarInfo() = default;
+
+ public:
+  static VarInfo Decode(const std::string &compact_labels, int info_code) {
```

Comment:
make info_code const?

Reply:
This is a style question. I prefer to not make function arguments that are passed by value `const`. Because the function has a copy of the data anyway, it cannot modify the value outside. So adding const doesn't really add anything and it just makes the function signature longer and a little less readable.

Reply:
I disagree on this point, it shows intent. Anytime you see const in the function signature you know it will be the same wherever you see it in the function body, but I leave it at your discretion, it is mostly preference.

Reply:
Fair point. I guess it boils down to stylistic preference